### PR TITLE
feat: txwy resource updater

### PR DESCRIPTION
**Pull request by Constrat**
@tsuk1ko This is the implementation of the repository, 
~but I noticed that not all files are update, for instance (the most important) [character_table.json](https://github.com/arkntools/arknights-toolbox-update/blob/data/tw/excel/character_table.json) does not have the new operators (From Ines --> today)~
I'm a dumbass, dont' mind me. I forgot txwy is in "the past" hence, doesn't have access to these operators yet.

### CODE REVIEW

**Comment by Tsuk1ko** [comment]
I was about to explain, lol

**Comment by Constrat** [comment]
> I was about to explain, lol

I was completely tunnel visioned on it because I was sure it was going to work but no changes were made so I was trying to explain to myself why xD

**Comment by Tsuk1ko** [comment]
I'm not sure if these files will suffice. If you want, I provide an [action](https://github.com/arkntools/arknights-toolbox-update/blob/master/.github/workflows/example.yml) to customize the download of certain excel files from a certain server. 
However, there will be some additional maintenance work in the future (mainly because each server requires different fbs).

Of course, you can also continue to pull it directly from my repo, but requires additional configuration of PAT in workflow.
If I haven't provide the files you need, let me know and I can add them.

**Comment by Constrat** [comment]
Thank you, but it seems like we have all we need (actually there are some json that we do not use)
We use:
- item_table.json
- building_data.json
- gacha_table.json
- character_table.json
- activity_table.json

**Comment by Tsuk1ko** [comment]
This is easy to solve, when I have free time (maybe on Saturday) I can change the folder structure to meet the requirements.

**Comment by Constrat** [comment]
> This is easy to solve, when I have free time (maybe on Saturday) I can change the folder structure to meet the requirements.

For now I've fixed it to compensate for this. (No hurry if you still want to it would be awesome because I can remove lots of control lines). If you can also manage to add `roguelike_topic_table.json` in the jsons the updater would go back as we used to have it 🙇🏻‍♂️

File: tools/ResourceUpdater/main.cpp [hunk 1]
```diff
@@ -87,158 +86,224 @@ int main([[maybe_unused]] int argc, char** argv)
...
+            std::cout << "------------Update roguelike replace for " << out << "------------" << std::endl;
+            if (!check_roguelike_replace_for_overseas(overseas_data_dir / in / "excel",
+                                                      resource_dir / "global" / out / "resource" / "tasks.json",
+                                                      official_data_dir / "gamedata" / "excel", cur_path / in)) {
+                std::cerr << "Update roguelike replace for overseas failed" << std::endl;
+                return -1;
+            }
+            else {
+                std::cout << "Done" << std::endl;
+            }
```
Comment (Constrat) [inline tools/ResourceUpdater/main.cpp:270]: IMPORTANT REF `roguelike_topic_table.json` MISSING FOR TXWY

**Comment by Constrat** [comment]
Also when this is merged in dev, we might need the repo to go public otherwise I think only me and @AnnAngela will be able to run the workflow.

**Comment by Tsuk1ko** [comment]
Sure, i will add it later.

**Comment by Constrat** [comment]
Currently waiting for:

- [ ] `roguelike_topic_table.json` for https://github.com/MaaAssistantArknights/MaaAssistantArknights/pull/7259#discussion_r1388465663.
- [ ] [arknights-toolbox-update](https://github.com/arkntools/arknights-toolbox-update) to go public for [actions/workflows/res-update-game](https://github.com/MaaAssistantArknights/MaaAssistantArknights/actions/workflows/res-update-game.yml).
- [ ] `/gamedata` in folder structure, to clean the code.

**Comment by AnnAngela** [comment]
> Also when this is merged in dev, we might need the repo to go public otherwise I think only me and @AnnAngela will be able to run the workflow.

maybe we can add a deploy key on that private repo to avoid that

**Comment by AnnAngela** [comment]
> Currently waiting for:
> 
> - [ ] `roguelike_topic_table.json` for https://github.com/MaaAssistantArknights/MaaAssistantArknights/pull/7259#discussion_r1388465663.
> - [ ] [arknights-toolbox-update](https://github.com/arkntools/arknights-toolbox-update) to go public for [actions/workflows/res-update-game](https://github.com/MaaAssistantArknights/MaaAssistantArknights/actions/workflows/res-update-game.yml).
> - [ ] `gamedata/` in [arknights-toolbox-update](https://github.com/arkntools/arknights-toolbox-update/tree/data) folder structure, to clean the code.

i can solve these today

**Comment by Tsuk1ko** [comment]
> > Also when this is merged in dev, we might need the repo to go public otherwise I think only me and @AnnAngela will be able to run the workflow.
> 
> maybe we can add a readonly token on that private repo to avoid that

Yes, making gamedata public is not such a good thing for me.

**Comment by Tsuk1ko** [comment]
@Constrat I have added `roguelike_topic_table.json` and `gamedata` folder.

I decided not to modify my action but to make a move before commit so it was simply done. XD

image

**Comment by AnnAngela** [comment]
I generated a token for tools, so I will start working this afternoon (UTC+8)

**Comment by AnnAngela** [comment]
@Constrat Hi, can you tell me what files should be downloaded and where to put them?

**Comment by Constrat** [comment]
- roguelike_topic_table.json MISSING
- item_table.json
- building_data.json
- gacha_table.json
- character_table.json
- activity_table.json

These are all the ones used (mainly from the official client repository) 
IIRC overseas and txwy do not use the building_data.json, also is there a way to download only the Taiwan data? To reduce the load?

**Comment by AnnAngela** [comment]
@Constrat We can use parameters to configure the tool to download the txwy data. So where should the data be placed?

**Comment by Constrat** [comment]
I had a method to directly clone the repository and check then create `arknights-toolbox-update` folder (in the ResourceUpdater folder. 
If could directly download it inside the `ArknightsGameData_Yostar` so that they all stay in the same directory (and the global_dirs map remains simple)

**Comment by AnnAngela** [comment]
@Constrat Hmmm, sorry for my poor English. I mean which path do you want me to download the data to?

**Comment by Constrat** [comment]
const auto yostar_data_dir = cur_path / "ArknightsGameData_YoStar";
Inside this.
cur_path is the project directory, (`./tools/ResourceUpdater/`)

Download in `./tools/ResourceUpdater/ArknightsGameData_Yostar`
Preferably download only from tw/gamedata/excel/files.json without the parent directory if possible

**Comment by Constrat** [comment]
Shouldn't the txwy download come before the ResourceUpdater.exe?

**Comment by Tsuk1ko** [comment]
@AnnAngela I'm thinking if it is better to clone my gamedata repo. 🤔

As I said before you will need to maintain fbs-head parameters if  using my action.

**Comment by Constrat** [comment]
Eh you're probably right. That way the resource updater only depends on one cpp file and there's no need to also change actions yml

**Comment by Tsuk1ko** [comment]
> The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)

Easy, I can provide a branch that only have tw data.

**Comment by Constrat** [comment]
> > The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)
> 
> Easy, I can provide a branch that only have tw data.

No that's not what I meant. Having only tw is still a +1, (lesser load, lesser traffic etc.) The issue remains with git. To clone it wants an empty folder. Hence I can't clone it in an existing repository, but that's what I would like to do xD

**Comment by Constrat** [comment]
> > The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)
> 
> Easy, I can provide a branch that only have tw data.
> 
> @Constrat What folder structure is best if I provide that? Just put all jsons in the root?

Keep the xxclient/gamedata/excel/jsons(so that it's identical to the official CN one)

**Comment by AnnAngela** [comment]
> > > The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)
> > 
> > 
> > Easy, I can provide a branch that only have tw data.
> 
> No that's not what I meant. Having only tw is still a +1, (lesser load, lesser traffic etc.) The issue remains with git. To clone it wants an empty folder. Hence I can't clone it in an existing repository, but that's what I would like to do xD
> 
> Maybe we can just directly repositories and start downloading directly from arktool's. I'll see what I can do at home (in 10hrs approx). @Tsuk1ko for now just create a new branch with only TW data, to have it ready when needed. Thank you :)

We can clone to temp folder and copy back to this repo

**Comment by Constrat** [comment]
... can't do much at uni without a pc xD

**Comment by Constrat** [comment]
> > > > The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)
> > > 
> > > 
> > > Easy, I can provide a branch that only have tw data.
> > 
> > 
> > No that's not what I meant. Having only tw is still a +1, (lesser load, lesser traffic etc.) The issue remains with git. To clone it wants an empty folder. Hence I can't clone it in an existing repository, but that's what I would like to do xD
> > Maybe we can just directly repositories and start downloading directly from arktool's. I'll see what I can do at home (in 10hrs approx). @Tsuk1ko for now just create a new branch with only TW data, to have it ready when needed. Thank you :)
> 
> We can clone to temp folder and copy back to this repo

But then updating the repo wouldn't work no? We would have to keep on cloning it every time overwriting everything.
During this weekend (finally) I'll see if it's feasible to just use arkntool's data, to completely remove this issue
The only comment I have is that it was a bit slower to update than Kengxiaoo ArknightsGameData_Yostar, it doesn't make that much of a difference but I would just like to know why is that

**Comment by AnnAngela** [comment]
> > > > > The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)
> > > > 
> > > > 
> > > > Easy, I can provide a branch that only have tw data.
> > > 
> > > 
> > > No that's not what I meant. Having only tw is still a +1, (lesser load, lesser traffic etc.) The issue remains with git. To clone it wants an empty folder. Hence I can't clone it in an existing repository, but that's what I would like to do xD
> > > Maybe we can just directly repositories and start downloading directly from arktool's. I'll see what I can do at home (in 10hrs approx). @Tsuk1ko for now just create a new branch with only TW data, to have it ready when needed. Thank you :)
> > 
> > 
> > We can clone to temp folder and copy back to this repo
> 
> But then updating the repo wouldn't work no? We would have to keep on cloning it every time overwriting everything. During this weekend (finally) I'll see if it's feasible to just use arkntool's data, to completely remove this issue The only comment I have is that it was a bit slower to update than Kengxiaoo ArknightsGameData_Yostar, it doesn't make that much of a difference but I would just like to know why is that

I believe the `git clone` will keep the timestamp so if files are not changed, the workflow can detect and cancel itself.

**Comment by Constrat** [comment]
> > > > > > The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)
> > > > > 
> > > > > 
> > > > > Easy, I can provide a branch that only have tw data.
> > > > 
> > > > 
> > > > No that's not what I meant. Having only tw is still a +1, (lesser load, lesser traffic etc.) The issue remains with git. To clone it wants an empty folder. Hence I can't clone it in an existing repository, but that's what I would like to do xD
> > > > Maybe we can just directly repositories and start downloading directly from arktool's. I'll see what I can do at home (in 10hrs approx). @Tsuk1ko for now just create a new branch with only TW data, to have it ready when needed. Thank you :)
> > > 
> > > 
> > > We can clone to temp folder and copy back to this repo
> > 
> > 
> > But then updating the repo wouldn't work no? We would have to keep on cloning it every time overwriting everything. During this weekend (finally) I'll see if it's feasible to just use arkntool's data, to completely remove this issue The only comment I have is that it was a bit slower to update than Kengxiaoo ArknightsGameData_Yostar, it doesn't make that much of a difference but I would just like to know why is that
> 
> I believe the `git clone` will keep the timestamp so if files are not changed, the workflow can detect and cancel itself.

That would be ideal, but as I said I want to see if we can directly use arkntool's.

**Comment by AnnAngela** [comment]
> > > > > > > The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)
> > > > > > 
> > > > > > 
> > > > > > Easy, I can provide a branch that only have tw data.
> > > > > 
> > > > > 
> > > > > No that's not what I meant. Having only tw is still a +1, (lesser load, lesser traffic etc.) The issue remains with git. To clone it wants an empty folder. Hence I can't clone it in an existing repository, but that's what I would like to do xD
> > > > > Maybe we can just directly repositories and start downloading directly from arktool's. I'll see what I can do at home (in 10hrs approx). @Tsuk1ko for now just create a new branch with only TW data, to have it ready when needed. Thank you :)
> > > > 
> > > > 
> > > > We can clone to temp folder and copy back to this repo
> > > 
> > > 
> > > But then updating the repo wouldn't work no? We would have to keep on cloning it every time overwriting everything. During this weekend (finally) I'll see if it's feasible to just use arkntool's data, to completely remove this issue The only comment I have is that it was a bit slower to update than Kengxiaoo ArknightsGameData_Yostar, it doesn't make that much of a difference but I would just like to know why is that
> > 
> > 
> > I believe the `git clone` will keep the timestamp so if files are not changed, the workflow can detect and cancel itself.
> 
> That would be ideal, but as I said I want to see if we can directly use arkntool's.

NP, so when the arkntool's data repo is ready, I will change the workflow to use that repo

**Comment by Constrat** [comment]
Is there a way to do it only with cpp? Without touching res-update yml?

**Comment by Tsuk1ko** [comment]
> Is there a way to do it only with cpp? Without touching res-update yml?

Read git token from env? Still need touch workflow because we must add the token.

**Comment by Tsuk1ko** [comment]
> > > The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)
> > 
> > 
> > Easy, I can provide a branch that only have tw data.
> > @Constrat What folder structure is best if I provide that? Just put all jsons in the root?
> 
> Keep it as-is with the xxclient/gamedata/excel/jsons(so that it's identical to the official CN one) With the new `gamedata/`

I means I can make a branch like this, just all files in root:

```
repo-root
├── xxx.json
├── yyy.json
└── zzz.json
```

Then you can clone it to where you want:

```bash
git clone REPO_URL --depth=1 xxx/ArknightsGameData_Yostar/zh_TW/gamedata/excel
```

Clone with checkout action is also fine.

But clone in cpp you need read the git token from env and pass it to command line?

**Comment by Constrat** [comment]
Oh, that makes sense. Make the Taiwan only branch like that. (Worst case we will use that)
Best case it won't be needed because we transfer to arkntool's full repo for overseas.

Regarding the token, we can just change the Res update yml to pass the token as parameter to the main function. It should be doable.

**Comment by AnnAngela** [comment]
@Constrat FYI, the action way is working: <https://github.com/AnnAngela/AnnAngela/actions/runs/6822851655/job/18555648270>

**Comment by Constrat** [comment]
I want to see first if I can do it with cpp as I have n idea.
(At least now I have it, Calculus II is taking a toll on me)

**Comment by AnnAngela** [comment]
> I want to see first if I can do it with cpp as I have n idea. (At least now I have it, Calculus II is taking a toll on me)

OK, I commented out the action code, and make the token as env `ARKNTOOLS_MAA_RESOURCE_UPDATER_TOKEN` so you can use it.

**Comment by Constrat** [comment]
Ok I've stolen a computer from a laboratory.

**Comment by Constrat** [comment]
> > > > The only small issue is that I don't think I can clone only the tw folder in the other directory of the Yostar repo (to have all the overseas files in only one place)
> > > 
> > > 
> > > Easy, I can provide a branch that only have tw data.
> > > @Constrat What folder structure is best if I provide that? Just put all jsons in the root?
> > 
> > 
> > Keep it as-is with the xxclient/gamedata/excel/jsons(so that it's identical to the official CN one) With the new `gamedata/`
> 
> I means I can make a branch like this, just all files in root:
> 
> ```
> repo-root
> ├── xxx.json
> ├── yyy.json
> └── zzz.json
> ```
> 
> Then you can clone it to where you want:
> 
> ```shell
> git clone REPO_URL -b BRANCH --depth=1 xxx/ArknightsGameData_Yostar/zh_TW/gamedata/excel
> ```
> 
> Clone with checkout action is also fine.
> 
> But clone in cpp you need read the git token from env and pass it to command line?

Ok make a branch like this, I would like to test if it slows / has issues (Should not happen no?)

**Comment by Tsuk1ko** [comment]
OK, I'll be free to make one in about an hour.

**Comment by Constrat** [comment]
@Tsuk1ko I don't think we can completely use arkntools repository for all overseas.
The character_table.json structure is completely different for: 
- EN
- JP
- KR
- and CN but I never wanted to switch for Official, too many responsabilited xD

Yours has a `key`, `value` structure, while the one we use and imeplement has only `keys `(char operator id) with subvalues (your `values`



</p>
</details>

**Comment by Tsuk1ko** [comment]
> @Tsuk1ko I don't think we can completely use arkntools repository for all overseas.
> 
> The character_table.json structure is completely different for: 
> 
> - EN
> 
> - JP
> 
> - KR
> 
> - and CN but I never wanted to switch for Official, too many responsabilited xD
> 
> 
> 
> Yours has a `key`, `value` structure, while the one we use and imeplement has only `keys `(char operator id) with subvalues (your `values`
> 
> 
> 
> 
> 
> 
> 
> </p>
> 
> </details> 

Those can be converted to dictionary, I will do it later.

**Comment by Constrat** [comment]
@AnnAngela I have no idea how to test actions in a "safe" environment with the resource updater changes.
Do you think this would work?

**Comment by Tsuk1ko** [comment]
@Constrat I finished. You can see if the current data is what you want.

The tw data only branch is `data-tw`.

**Comment by Constrat** [comment]
I just left the lab :( . I'll have a look at it this late afternoon, I'll have the entire weekend free

**Comment by Constrat** [comment]
Ok, I've changed the github actions.
I'm merging this as I don't have any other way to test, if anything goes wrong I'll just un merge the changes.
This is not MaaCore / WpgGui breaking, it's just CI.

**Comment by Constrat** [comment]
```bash
------- Clone arknights-toolbox-update for Taiwan -------
Cloning into 'D:\a\MaaAssistantArknights\MaaAssistantArknights\tools\ResourceUpdater\x64\Release\ArknightsGameData_YoStar\zh_TW\gamedata\excel'...
remote: Write access to repository not granted.
fatal: unable to access 'https://github.com/arkntools/arknights-toolbox-update.git/': The requested URL returned error: 403
git cmd failed
Error: Process completed with exit code 1.
```
@Tsuk1ko Well, we are almost there. In the end my not wanting to split the workload between Actions and cpp is not paying off.
Why would it request write access to clone a repository?

**Comment by Tsuk1ko** [comment]
> ```shell
> ------- Clone arknights-toolbox-update for Taiwan -------
> Cloning into 'D:\a\MaaAssistantArknights\MaaAssistantArknights\tools\ResourceUpdater\x64\Release\ArknightsGameData_YoStar\zh_TW\gamedata\excel'...
> remote: Write access to repository not granted.
> fatal: unable to access 'https://github.com/arkntools/arknights-toolbox-update.git/': The requested URL returned error: 403
> git cmd failed
> Error: Process completed with exit code 1.
> ```
> 
> @Tsuk1ko Well, we are almost there. In the end my not wanting to split the workload between Actions and cpp is not paying off. Why would it request write access to clone a repository? The cpp syntax seems to be correct, because it directly gives an error (doesn't ask for a password like this case)
> 
> Testing on my machine through powershell
> @MistEO the changes I've made are in fact, not working. instead of reverting the entire branch, I'm just pushing on `dev` a fix that disables the txwy updater, until we can fix this permission issue.

@AnnAngela Did you select this to Read-only? I have tested it works.

image

@Constrat And I'm curious if you used the in token parameter.

https://github.com/MaaAssistantArknights/MaaAssistantArknights/blob/57e73df3d4294d7335d558474374c0c33db93793/.github/workflows/res-update-game.yml#L35-L39

I didn't see you use this (may get by `argv[1]` i think?). And you need clone my repo like this:

```bash
git clone https://TOKEN@github.com/arkntools/arknights-toolbox-update.git -b data-tw
```

I guess that git in the workflow environment may have a workflow token by default (`${{ github.token }}` in action context), and it only has read-only access to current repo, so it reported a permission error.

**Comment by Constrat** [comment]
It's not in this PR, after I've merged I changed that

**Comment by Tsuk1ko** [comment]
@Constrat Are you sure you can get the token from the parameter? Have you verified the value?

**Comment by Tsuk1ko** [comment]
🤔 You can try to output this. If it is right, it should be printed as a bunch of asterisks. GitHub actions will automatically block all secrets output.

https://github.com/MaaAssistantArknights/MaaAssistantArknights/blob/9ce9139cee54bc821afa25b6a14ebd74f6f8c5c4/tools/ResourceUpdater/main.cpp#L77C28-L77C42

**Comment by Constrat** [comment]
Wait I think I already tried this yesterday and it showed some asterisks, but only 3 or 4. (Does this mean something?)

**Comment by Tsuk1ko** [comment]
I suggest you troubleshoot like this:

1. get token from env directly (cpp should be able to do it, `getenv()` or sth)
2. if not work, try generate a PAT by yourself

Generate setp:

1. Go to your "fine-grained personal access tokens" settings -> https://github.com/settings/tokens?type=beta
2. Follow these setting:
image
image
4. Generate, and test it locally

If it works, replace current secrets `ARKNTOOLS_MAA_RESOURCE_UPDATER` and action should work.

**Comment by Constrat** [comment]
Ok I think I know what's the problem: @AnnAngela created the token, but I think the repository was just the `data-downloader` repository, and the `arknights-toolbox-update`, this could probably the reason why it gives an error.

**Comment by AnnAngela** [comment]
> Ok I think I know what's the problem: @AnnAngela created the token, but I think the repository was just the `data-downloader` repository, and the `arknights-toolbox-update`, this could probably the reason why it gives an error.
> 
> I was now trying to use the GitHub actions workflow file to clone the repositories directly from there and not from the cpp file (hoping this could solve the problem) but I think it will just be the same because the token is not for the correct repository. As I said before I don't think I have the permission to use my token with MAA organization (I have no idea what I should do to fix this)

the token that I generated has the permission to clone these repo:
- arkntools/arknights-toolbox-data
- arkntools/data-downloader

does arknights-toolbox-update be needed?

**Comment by Tsuk1ko** [comment]
> > Ok I think I know what's the problem: @AnnAngela created the token, but I think the repository was just the `data-downloader` repository, and the `arknights-toolbox-update`, this could probably the reason why it gives an error.
> > 
> > I was now trying to use the GitHub actions workflow file to clone the repositories directly from there and not from the cpp file (hoping this could solve the problem) but I think it will just be the same because the token is not for the correct repository. As I said before I don't think I have the permission to use my token with MAA organization (I have no idea what I should do to fix this)
> 
> the token that I generated has the permission to clone these repo:
> - arkntools/arknights-toolbox-data
> - arkntools/data-downloader
> 
> does arknights-toolbox-update be needed?

Yes... Actually the data is in arknights-toolbox-update

**Comment by Constrat** [comment]
XD so it was this all along.

**Comment by Constrat** [comment]
Well, since I was already working  on the actions I might as well continue and transfer the cloning and git actions to the yml (if it will work xD)

**Comment by Tsuk1ko** [comment]
>  As I said before I don't think I have the permission to use my token with MAA organization (I have no idea what I should do to fix this)

@Constrat You means you don't have permission to set secrets?

**Comment by Constrat** [comment]
No worries, when AnnAngela adds toolbox-update to the pat nothing else matters and I should be able to close.

**Comment by Constrat** [comment]
> > As I said before I don't think I have the permission to use my token with MAA organization (I have no idea what I should do to fix this)
> 
> @Constrat You means you don't have permission to set secrets?

For clearing up. If I generate my PAT, when coding in Maaassistantarknights/maaassistantarknights I can't use mine (intellicode doesn't show up) but nonworries,.

**Comment by AnnAngela** [comment]
> No worries, when AnnAngela adds toolbox-update to the pat nothing else matters and I should be able to close.

done, the token has been updated. you can use it to clone arknights-toolbox-update

**Comment by Tsuk1ko** [comment]
@Constrat This can be removed I think.

https://github.com/MaaAssistantArknights/MaaAssistantArknights/blob/64fbe3052d61d2a6ad41fda4cf43f5acc34311a0/tools/ResourceUpdater/main.cpp#L78

**Comment by Constrat** [comment]
Let me finish break fast. XD I spent the entire morning understanding why it wasn't working.
In two hours everything should be resolved

**Comment by Constrat** [comment]
> > As I said before I don't think I have the permission to use my token with MAA organization (I have no idea what I should do to fix this)
> 
> @Constrat You means you don't have permission to set secrets?

Btw I found what I would have needed (using my key in the organization) I didnt' dig deep enough.

**Comment by Constrat** [comment]
I don't know, cloning with actions seems much slower then cloning with cpp, @AnnAngela what do you think is best?

**Comment by Tsuk1ko** [comment]
Yeah, I think it should be better. Just half a minute slower is not a big deal.

**Comment by Constrat** [comment]
> Yeah, I think it should be better. Just half a minute slower is not a big deal.
> 
> There is no need to change the cpp when switching data source in the future,

If only that was true XD.
Trust me, if anything breaks we will ALWAYS be changing the cpp

**Comment by Constrat** [comment]
I think I can understand why everything was done in the cpp. That way you can easily work on local, instead of relying on github actions. (aka setupping actions in your own repository etc etc, needing the token variable)

**Comment by Tsuk1ko** [comment]
> I think I can understand why everything was done in the cpp. That way you can easily work on local, instead of relying on github actions. (aka setupping actions in your own repository etc etc, needing the token variable)

Well, I think separate them out is more convenient on local. Since git clone is a one-time operation, you can just clone the repos into local once and run cpp multiple times without clearing the git files and waiting for clone.

**Comment by Constrat** [comment]
I didn't think about that, I might create a small py script to allow Devs to easily clone them (with python you should be able to give GitHub secrets (I hope))

**Comment by AnnAngela** [comment]
> I don't know, cloning with actions seems much slower then cloning with cpp, @AnnAngela what do you think is best?
> (I'm working on my [Constrat/MaaAssistantArknights](https://github.com/Constrat/MaaAssistantArknights))
> 
> Technically (the latest commit in my repo) should be the cleanest way. Github actions works with github and the .exe (cpp) does all the other things such as modifying jsons (while only reading the folders, not as repositories)

idk why, maybe because the action way is running on nodejs?

I don't think there is a significant difference between action way and cpp way. FYI IDK how to use cpp, so I can't help when using cpp way.

**Comment by Constrat** [comment]
Thank you! I'm going with actions because I think I can add a filtered checkout (to exponentially decrease time)

**Comment by AnnAngela** [comment]
@Constrat hi bro, can you verify that the workflow can update txwy data properly and the txwy data can be used properly? if so, we can remake the  announcement #7165 to tell the users this news

**Comment by Constrat** [comment]
### IMPORTANT
@AnnAngela @MistEO the Actions works as intended. Txwy should receive the data like all the other clients as usual. The performance has been increased because in the action for ALL clients only the necessary jsons are used. Later I will add a list of the used jsons.
No new content has been released yet so I can't really be sure if everything works, but in my local environment it was working.

**Pull request merged**